README
At present this version of EasyGP is only intended for developers
though you may try it if you like - but it is not at a point where
you will be able keep your medical records accurately or safely so
should not on any account be used for that purpose.

Main developers:

Dr. Richard Terry <rterry@pacific.net.au>
Dr. Ian Haywood <ihaywood3@gmail.com>

**************
pre-requisites
**************
postgresql, version 8.4 or greater
gambas, version 3.0 (i.e. the developement svn branch)
ruby (if you want to compile the docs)

*************
quick-install
*************
1) Download the svn to a directory of your choice:

 user ~  $ mkdir /home/user/easygp
 user ~  $ cd /home/user/easygp
 user ~  easygp$ svn checkout svn://ozdocit.org/easygp/trunk
 
 you will see a whole lot of files checked out for example:
 
A    trunk/debian/easygp-server.install                                                                                                                                  
A    trunk/debian/changelog                                                                                                                                              
A    trunk/debian/copyright                                                                                                                                              
A    trunk/debian/easygp-client.install                                                                                                                                  
A    trunk/debian/make-server.sh        

Checked out revision xyz.

2) Compile the help files, these can only be run from within EasyGP at
   the current time.
   
 user ~ easygp  $  $  cd trunk/client/help/
 user ~ easygp/trunk/client/help  $  ruby basicdoc.rb easygp.bd
 user ~ easygp/trunk/client/help  $
 
 3) Refer to db/README for instriuctions on how to set up the database

 4) Run Gambas3 and select your project as the client/ directory under this directory,
    and from within the IDE make sure that the startup file is
    client/startup/modStartup.module (which should be the default), 
    then run the project

 5)  Log in with the user and passsword you set when creating the database
     back in step 3), the client should now start a wizard to allow you 
     to set the practice organisation
 

     
   
 


debian/changelog
easygp (0.529svn2462-8xenial3) xenial; urgency=medium

  * fix compilation with latest gambas

 -- Ian Haywood <ian@haywood.id.au>  Mon, 26 Jun 2017 21:55:17 +1000

easygp (0.529svn2462-5) zesty; urgency=medium

  * fix dependency gambas3-gb-web -gb-media -gb-media-form
  * take out v4l support: just too old

 -- Ian Haywood <ian@haywood.id.au>  Fri, 09 Jun 2017 07:32:55 +1000

easygp (0.529svn2462) zesty; urgency=medium

  * new svn version, major changes
  * try versioning of dependencies
  * major update from svn
  * going back to launchpad

 -- Ian Haywood <ian@haywood.id.au>  Thu, 08 Jun 2017 12:55:15 +1000

easygp (0.401svn2002) easygp-latest; urgency=low

  * Latest version

 -- Ian Haywood <ian@haywood.id.au>  Wed, 01 Jul 2015 13:57:42 +1000

easygp (0.376svn1877) precise; urgency=low

  * New version from weekend Dorrigo
  * bugfixes to item lists
  * fixes to FRecordVisit

 -- Ian Haywood <ihaywood@ozdocit.org>  Wed, 10 Sep 2014 20:23:18 +1000

easygp (0.376svn1866local2) precise; urgency=low

  * new version
  * add mplayer, aspell to dependencies
  * local mods for ABNs
  * local mods for handling JPEGs

 -- Ian Haywood <ian@haywood.id.au>  Thu, 15 May 2014 21:58:18 +1000

easygp (0.359svn1830) precise; urgency=low

  * fix dependencies: add gb.v4l and gb.web
  * Database data restored
  * fix path functions and default paths
  * add highlight component dependency
  * fixed ians in db dump (again)
  * more dependencies, remove function from easygp-server.postinst
  * new dependencies

 -- Ian Haywood <ian@haywood.id.au>  Tue, 25 Feb 2014 22:37:04 +1100

easygp (0.288svn1516) precise; urgency=low

  * database formatting fixes

 -- Ian Haywood <ian@haywood.id.au>  Mon, 10 Jun 2013 10:27:39 +1000

easygp (0.286svn1497) precise; urgency=low

  * fix ownership in database

 -- Ian Haywood <ian@haywood.id.au>  Fri, 07 Jun 2013 16:45:58 +1000

easygp (0.286svn1486) precise; urgency=low

  * make sure compile gambas binary

 -- Ian Haywood <ian@haywood.id.au>  Fri, 31 May 2013 22:22:31 +1000

easygp (0.286svn1474) precise; urgency=low

  * another test release for Richard

 -- Ian Haywood <ian@haywood.id.au>  Thu, 30 May 2013 19:17:23 +1000

easygp (0.270svn1428) precise; urgency=low

  * fixes to client for admin mode
  * use the right versioning system

 -- Ian Haywood <ian@haywood.id.au>  Sun, 21 Apr 2013 18:12:37 +1000

easygp (0.8.1) precise; urgency=low
  * fixes to permissions
 -- Ian Haywood <ian@haywood.id.au>  Sat, 20 Apr 2013 18:46:34 +1000

easygp (0.8) precise; urgency=low

  * Update to billing
  * database update, testing install again

 -- Ian Haywood <ian@haywood.id.au>  Sat, 20 Apr 2013 16:46:34 +1000

easygp (0.7) precise; urgency=low

  * split easygp-client to easygp-doc

 -- Ian Haywood <ian@haywood.id.au>  Fri, 28 Sep 2012 11:52:15 +1000

easygp (0.5) precise; urgency=low

  * New attempt at working version

 -- Ian Haywood <ian@haywood.id.au>  Sat, 15 Sep 2012 13:11:13 +1000

easygp (0.4~lucid2) lucid; urgency=low

  * fixes to server install

 -- Ian Haywood <ian@haywood.id.au>  Mon, 03 Jan 2011 19:32:51 +1000

easygp (0.4~lucid1) lucid; urgency=low

  * extended server install

 -- Ian Haywood <ihaywood@gnu.org>  Mon, 03 Jan 2011 19:07:45 +1000

easygp (0.3~lucid8) lucid; urgency=low

  * added more dependencies
  * removed superfluous dependency on gb.web

 -- Ian Haywood <ihaywood@gnu.org>  Sun, 21 Nov 2010 10:45:07 +1000


easygp (0.3~lucid6) lucid; urgency=low

  * remove superfluous dependency on gb.db.form

 -- Ian Haywood <ihaywood@gnu.org>  Sun, 21 Nov 2010 02:32:01 +1000

easygp (0.3~lucid5) lucid; urgency=low

  * add dependency for gb.crypt correctly

 -- Ian Haywood <ihaywood@gnu.org>  Sat, 20 Nov 2010 22:45:32 +1000

easygp (0.3~lucid4) lucid; urgency=low

  * add dependency for gb.crypt

 -- Ian Haywood <ihaywood@gnu.org>  Sat, 20 Nov 2010 22:07:32 +1000

easygp (0.3~lucid3) lucid; urgency=low

  * remove use of gb.chart

 -- Ian Haywood <ihaywood@gnu.org>  Sat, 20 Nov 2010 20:17:37 +1000

easygp (0.3~lucid2) lucid; urgency=low

  * Fix build-dependencies

 -- Ian Haywood <ihaywood@gnu.org>  Sat, 20 Nov 2010 17:16:31 +1000

easygp (0.2~lucid1) lucid; urgency=low

  * Completed man pages for database admin
  * svn revision to 351

 -- Ian Haywood <ihaywood@gnu.org>  Sat, 20 Nov 2010 15:48:31 +1000


easygp (0.1) unstable; urgency=low

  * Initial release

 -- Ian Haywood <ihaywood@gnu.org>  Thu, 10 Oct 2010 23:11:32 +1000


help/basicdoc.rb
#!/usr/bin/ruby

$output = nil
$section = nil
$subsection = nil
$subsubsection = nil
$subsubsubsection = nil
$list = false
$pre = false
                  # from the 200 commonest words in English
$common_words = ["and","but","can","will","should","the","you","was","are","they","one","his",
                 "have","this","from","some","what","out","other","were","then","all","there",
                 "where","word","how","said","each","she","which","way","about","many","then",
                 "them","would","like","these","her","thing","see","two","him","has","more",
                 "could","come","did","most","who","over","know","than","first","may","down",
                 "side","been","any","new","find","take","place","get","live","after","back",
                 "little","big","only","came","every","give","our","under","very","few","just",
                 "much","great","say","before","turn","because","same","differ","right","left",
                 "too","does","set","well","also","small","end","put","must","high","such","why",
                 "ask","change","went","kin","off","need","try","again","near","self",
                 # words that trial-an-error so need to be excluded
                 "don","sub","here",
                 # words common in computer documentation
                 "click","mouse","key","window","screen"]


def indexise(str)
  str.downcase!
  str.split(/[^a-z]/).each do |word|
    if word.length > 2 and not $common_words.include?(word)
      loc = $subsubsection
      loc = $subsection unless loc
      loc = $section unless loc
      $index.write("%s\t%s\n" % [word,loc])
    end
  end
end

def make_key(name)
  r = name.downcase
  r.strip!
  r.delete!("\"':;,.?\\|[]{}<>()*&%$@!`~\t")
  r.gsub!(' ','-')
  return r
end

def parse_name(name)
  if name =~ /([a-z\/]+\.(png|jpg|jpeg)) (.+)/
    title = $3
    file = $1
  else
    file = nil
    title = name
  end
  return [title,file]
end

def new_html_file(key,title,file)
  if $output
    $output.write("\n<!--FTR--></body></html>")
    $output.close
  end
  title = esc(title)
  if $debian
    $output = File.new("build/"+key+".html",'w')
  else
    $output = File.new(key+".html",'w')
  end
  $output.write("<html>\n<head><title>%s</title>\n<!--HDR-->\n</head><body>\n" % title)
  if file
    $output.write("<img src=\"%s\" />\n" % file)
  end
  $output.write("<h2>%s</h2>\n" % title)
  indexise(title)
end

def esc(str)
  s2 = str.gsub('&','&amp;')
  s2 = s2.gsub('<','&lt;')
  s2 = s2.gsub('>','&gt;')
  return s2.gsub("\t"," ")
end


def latex_esc(str)
  s2 = str.gsub('\\', '\\\\')
  s2.gsub!('/','\\slash ')
  s2.gsub!('$', '\\$')
  s2.gsub!('%', '\\%')
  s2.gsub!('&', '\\&')
  s2.gsub!('<', '$<$')
  s2.gsub!('>', "$>$")
  return s2.gsub('_', '\\textunderscore ')
end

def toc_add(node,title,up)
  $toc.write("%s\t%s\t%s\n" % [node,title,up])
  $toctable += [{:node=>node,:title=>title,:up=>up}]
end

def process_file(fd)
  fd.each_line  do |line|
    case line
    when /^\.include (.*)/
      begin
        $latex.write(latex_esc("\n\n\\texttt{[File %s]}\n\n" % $1))
        process_file(File.new($1))
      rescue
	$latex.write(latex_esc("\n\n\\textt{[No such file %s]}\n\n" % $1))
        $output.write("<p><tt>No such file %s</tt></p>" % $1)
        $stderr.write("no such file %s\n" % $1)
      end
    when /^\.section (.*)/
      title, file = parse_name($1)
      key = make_key(title)
      $section = key
      $subsection = nil
      $subsubsection = nil
      toc_add(key,title,"root")
      $latex.write("\\chapter{%s}\n\\label{%s}\n" % [latex_esc(title),key])
      new_html_file(key,title,file)
    when /^\.subsection (.*)/
      title, file = parse_name($1)
      key = make_key(title)
      $subsection = key 
      $subsubsection = nil
      toc_add(key,title,$section)
      $latex.write("\\section{%s}\n\\label{%s}\n" % [latex_esc(title),key])
      new_html_file(key,title,file)
    when /^\.subsubsection (.*)/
      title, file = parse_name($1)
      key = make_key(title)
      $subsubsection = key
      toc_add(key,title,$subsection)
      $latex.write("\\subsection{%s}\n\\label{%s}\n" % [latex_esc(title),key])
      new_html_file(key,title,file)
    when /^\.paragraph (.*)/,/^\.subsubsubsection (.*)/
      $output.write("\n<h3>%s</h3>\n" % esc($1))
      $latex.write("\\subsubsection{%s}\n" % latex_esc($1))
      indexise($1)
    when /^\.link ([^ ]+) ?(.*)/
      if $2.nil? or $2 == ''
        $output.write("<a href=\"%s\">%s</a>" % [$1,$1])
        $latex.write("{\\tt %s}" % latex_esc($1))
      else
        $latex.write("%s ({\\tt %s})" % [latex_esc($2),latex_esc($1)])
        $output.write("<a href=\"%s\">%s</a>" % [$1,$2])
      end
    when /^\.xref (.*)/
      $output.write("<a href=\"%s.html\">%s</a>" % [make_key($1),$1])
      $latex.write("%s (see page \\pageref{%s})" % [$1,make_key($1)])
    when /^\.bold (.*)/
      $output.write("<b>%s</b> " % esc($1))
      $latex.write("{\\bf %s} " % latex_esc($1))
      indexise($1)
    when /^\.italic (.*)/
      $output.write("<i>%s</i> " % esc($1))
      $latex.write("\\emph{%s} " % latex_esc($1))
      indexise($1)
    when /^\.image ([^ ]+) ?(.*)/
      fname = $1.strip
      $output.write("<img src=\"%s\" />" % fname)
      if File.exist?(fname)
	if $2.nil? or $2 == ''
          width = 0.8
        else
          width = $2
        end 
        $latex.write("\\includegraphics[width=%s\\textwidth]{%s}\n" % [width,fname])
      else
        $stderr.write("no such file %s\n" % fname)
        $latex.write("\n\n{\\tt File %s does not exist}\n\n" % latex_esc(fname))
      end 
    when /^\.item(.*)/
      unless $list
        $output.write("\n<ul>\n")
        $latex.write("\\begin{itemize}")
        $list = true
      end
      $output.write("<li>%s" % esc($1))
      $latex.write("\n\\item %s" % latex_esc($1))
      indexise($1)
    when /^\.imagetext ([^ ]+) ?(.*)/
      fname = $1.strip
      if $2.nil? or $2 == ''
        width = "0.3"
        swidth = "0.65"
      else
        width = $2
	swidth = (1-$2.to_f)-0.05
      end
      $output.write("<table><tr><td><img src=\"%s\" /></td><td>" % fname)
      if File.exist?(fname)
        fname = "\\includegraphics[width=\\linewidth]{%s}" % fname
      else
        $stderr.write("File does not exist %s\n" % fname)
        fname = "{\\tt File %s does not exist}" % latex_esc(fname)
      end
      frag = <<EOF
\\begin{figure}[ht] 
\\begin{minipage}[b]{%s\\linewidth} 
\\centering %s
\\end{minipage}
\\hspace{0.5cm}
\\begin{minipage}[b]{%s\\linewidth}
EOF
        $latex.write(frag % [width,fname,swidth])
    when /^\.end imagetext/
      if $list
        $output.write("</ul>")
        $latex.write("\n\\end{itemize}\n")
        $list = false
      end
      $output.write("</td></tr></table>")
      $latex.write("\n\\end{minipage}\n\\end{figure}\n")
    when '',"\n"
      if $list 
        $output.write("\n</ul>\n")
        $latex.write("\\end{itemize}\n")
        $list = false
      else
        $output.write("<p/>\n")
        $latex.write("\n\n")
      end
    when /\.centre (.*)/
      $output.write("<p align=\"center\">%s</p>" % esc($1))
      $latex.write("\\begin{center}\n%s\n\\end{center}\n" % latex_esc($1))
      indexise($1)
    when /\.pre$/
      $output.write("<pre>\n")
      $latex.write("\\begin{verbatim}\n")
      $pre = true
    when /\.end pre/
      $output.write("\n</pre>\n")
      $latex.write("\\end{verbatim}\n")
      $pre = false
    when /\.latex (.*)/
      $latex.write($1+"\n")
    when /\.table(.*)/
      $output.write("<table%s><tr><td>" % $1)
      $latex.write("\\begin{tabular}") # then .latex for specific commands
    when /\.col/
      $output.write("</td><td>")
      $latex.write(" & ")
    when /\.row/
      $output.write("</td></tr><tr><td>\n")
      $latex.write("\\\\\n")
    when /\.end table/
      $output.write("</td></tr></table>\n")
      $latex.write("\\end{tabular}\n")
    else
      if $pre
        $latex.write(line) 
        $output.write(line)
      else
        $latex.write(latex_esc(line))
        $output.write(esc(line))
      end
      indexise(line)
    end
  end
  fd.close
end

if ARGV[0] == "-h" or ARGV.length == 0
  STDOUT.write(<<EOF
Basicdoc v0.1

Usage: ruby basicdoc.rb filename
EOF
)
else
  fd = File.new(ARGV[0])
  $latex = File.new('easygp.tex','w')
  $latex.write File.open("intro.tex") {|f| f.read }
  $toctable = []
  if ARGV[1] == '--debian'
    system 'mkdir -p build'
    $debian = true
    $index = File.new('build/index.txt.in','w')
    $toc = File.new('build/toc.txt','w')
  else
    $debian = false
    $index = File.new('index.txt.in','w')
    $toc = File.new('toc.txt','w')
  end
 
  process_file(fd)
  $output.write("\n<!--FTR-->\n</body></html>\n")
  $output.close
  $toc.close
  $index.close
  Dir.chdir("build") if $debian
  system "sort -u < index.txt.in > index.txt"
  system "rm index.txt.in"
  if $debian
    Dir.chdir("..")
    system "find -type d \\( \\! -path \"*.svn*\" \\) \\( \\! -path \"*build*\" \\) -exec mkdir -p build/\\{} \\;"
    system "find -name '*.png' \\! -path '*build*' -exec cp \\{} build/\\{} \\;"
  end
  $latex.write File.open("gpl-3.0.tex") {|f| f.read}
  $latex.write("\n\\end{document}\n")
  $latex.close
  #system "pdflatex easygp.tex"
end
